### PR TITLE
Change Classroom to Room Data Model

### PR DESCRIPTION
**Pull request by AndrewDPham**
Change classrooms to rooms. Have yet to add LESSONS and TAGS to classrooms. Added user to get their specific class (that they created) functionality.

### CODE REVIEW

File: capstone/src/main/webapp/index.html [hunk 1]
```diff
@@ -8,7 +8,7 @@
     <link rel="stylesheet" href="style.css">
     <script src="script.js"></script>
     <script src="https://apis.google.com/js/platform.js" async defer></script>
-    <meta name="google-signin-client_id" content=''>
+    <meta name="google-signin-client_id" content='1002281613740-gsmtp3tteqgtoformikuglhu8fr388et.apps.googleusercontent.com'>
```
Comment (kyleasmith1) [inline capstone/src/main/webapp/index.html]: Careful

File: capstone/src/main/webapp/login.html [hunk 2]
```diff
@@ -8,7 +8,7 @@
     <link rel="stylesheet" href="style.css">
     <script src="script.js"></script>
     <script src="https://apis.google.com/js/platform.js" async defer></script>
-    <meta name="google-signin-client_id" content='914921573408-h9cb4rja119k5c92jaa669lvn3q4fvlc.apps.googleusercontent.com'>
+    <meta name="google-signin-client_id" content='1002281613740-gsmtp3tteqgtoformikuglhu8fr388et.apps.googleusercontent.com'>
```
Comment (kyleasmith1) [inline capstone/src/main/webapp/login.html]: Here too

File: capstone/src/main/webapp/index.html (hunk 1 above)
Comment (AndrewDPham) [reply capstone/src/main/webapp/index.html]: Sorry I forgot to repush

File: capstone/src/main/webapp/login.html (hunk 2 above)
Comment (AndrewDPham) [reply capstone/src/main/webapp/login.html]: Got it

File: capstone/src/main/java/com/google/sps/data/Room.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    public Room(Entity entity) {
+        this.entity = entity;
+    }
+
+    public Room(User host, String title, String description) { 
+        this.entity = new Entity(Room.ROOM_ENTITY_NAME);
+        this.entity.setProperty(Room.TITLE_PROPERTY_KEY, title);
+        this.entity.setProperty(Room.DESCRIPTION_PROPERTY_KEY, description);
+        this.entity.setProperty(Room.HOST_PROPERTY_KEY, host.getUserKey());
+        this.entity.setProperty(Room.FOLLOWERS_PROPERTY_KEY, new ArrayList<>());
```
Comment (kyleasmith1) [inline capstone/src/main/java/com/google/sps/data/Room.java]: You don't need to set the "new ArrayList<>()" properties because datastore just recognizes them as null anyways. Instead in getter/setter methods check if those values are null and initialize them there.

File: capstone/src/main/java/com/google/sps/data/Room.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        return (String) this.entity.getProperty(Room.TITLE_PROPERTY_KEY);
+    }
+
+    public void setTitle(String title){
+        this.entity.setProperty(Room.TITLE_PROPERTY_KEY, title);
+    }
+
+    @SuppressWarnings("unchecked")
+    public List<Key> getAllFollowers() {
+        return (ArrayList<Key>) this.entity.getProperty(Room.FOLLOWERS_PROPERTY_KEY);
```
Comment (kyleasmith1) [inline capstone/src/main/java/com/google/sps/data/Room.java:55]: Check if this property is null

File: capstone/src/main/java/com/google/sps/data/Room.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    @SuppressWarnings("unchecked")
+    public List<Key> getAllForms() { 
+        return (ArrayList<Key>) this.entity.getProperty(Room.FORMS_PROPERTY_KEY); 
+    }
+    
+    @SuppressWarnings("unchecked")
+    public void addStudent(User follower) {
+        List<Key> followers = (ArrayList<Key>) this.entity.getProperty(Room.FOLLOWERS_PROPERTY_KEY);
```
Comment (kyleasmith1) [inline capstone/src/main/java/com/google/sps/data/Room.java]: Before you add the students to the followers list you need to check if the property is null

File: capstone/src/main/java/com/google/sps/data/Room.java (hunk 3 above)
Comment (AndrewDPham) [reply capstone/src/main/java/com/google/sps/data/Room.java]: Sounds good, will fix this in this PR.

File: capstone/src/main/java/com/google/sps/data/Room.java (hunk 5 above)
Comment (AndrewDPham) [reply capstone/src/main/java/com/google/sps/data/Room.java]: Sounds good. Also I forgot I have it as "addStudent".

**Review by mattlgy — APPROVED** [review]

File: capstone/src/main/java/com/google/sps/data/Room.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        return (ArrayList<Key>) this.entity.getProperty(Room.FORMS_PROPERTY_KEY); 
+    }
+    
+    @SuppressWarnings("unchecked")
+    public void addFollower(User follower) {
+        List<Key> followers = (ArrayList<Key>) this.entity.getProperty(Room.FOLLOWERS_PROPERTY_KEY);
+        if(followers == null){
+            followers = new ArrayList<>();
+        }
+        followers.add(follower.getUserKey());
```
Comment (707090) [inline capstone/src/main/java/com/google/sps/data/Room.java:63]: In the case that it was null, it sets followers to a new list, but that new list is not actually on the entity, so the new list with one item in it just gets dropped on the floor after this function exits. Add a test to cover this case (adding the first follower to a Room created with a preexisting entity that does not have followers)

File: capstone/src/main/java/com/google/sps/data/Room.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    @SuppressWarnings("unchecked")
+    public void removeForm(Form form) {
+        List<Key> forms = (ArrayList<Key>) this.entity.getProperty(Room.FORMS_PROPERTY_KEY);
+        forms.remove(form.getFormKey());
+    }
+
+    @SuppressWarnings("unchecked")
+    public boolean isFollowerInClass(User follower) {
```
Comment (707090) [inline capstone/src/main/java/com/google/sps/data/Room.java]: In "class"?

File: capstone/src/main/java/com/google/sps/servlets/DashboardHandlerServlet.java [hunk 8]
```diff
@@ -27,31 +31,28 @@
     @Override
     public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-        
-        // TODO: In a later CL we will make it so that it only gets User-specific classrooms
-        PreparedQuery results = datastore.prepare(new Query(Classroom.CLASSROOM_ENTITY_NAME));
+        Key userKey = ((User) request.getAttribute(User.USER_ENTITY_NAME)).getUserKey();
+        Filter roomFilter = new FilterPredicate(Room.HOST_PROPERTY_KEY, FilterOperator.EQUAL, userKey);
+        Query query = new Query(Room.ROOM_ENTITY_NAME).setFilter(roomFilter);
+        PreparedQuery results = datastore.prepare(query);
```
Comment (707090) [inline capstone/src/main/java/com/google/sps/servlets/DashboardHandlerServlet.java:37]: This change should probably be part of another PR. The logical group of changes this PR has (changing the naming/keys/etc for Classroom) is unrelated to this change, so they should be separate PRs. This could even be just a super duper simple 4 line PR which adds this feature

File: capstone/src/main/webapp/dashboard-loader.js [hunk 9]
```diff
@@ -1,44 +1,38 @@
...
-window.addEventListener('authorized', getClassrooms);
+window.addEventListener('authorized', getRooms);
 
-function createClassroom(subject) {
-    var auth2 = gapi.auth2.getAuthInstance();
-    var name = "";
-    var email = "";
+function createRoom(title, description) {
+    console.log("Title:" + title);
+    console.log("Description:" + description);
```
Comment (707090) [inline capstone/src/main/webapp/dashboard-loader.js]: Remember not to commit debug lines

**Review by 707090 — CHANGES_REQUESTED** [review]

**Comment by 707090** [comment]
Also its useful to mark conversations resolved once they are not longer relevant, just so I dont need to read them checking if they already said something I want to say

File: capstone/src/main/java/com/google/sps/servlets/DashboardHandlerServlet.java (hunk 8 above)
Comment (AndrewDPham) [reply capstone/src/main/java/com/google/sps/servlets/DashboardHandlerServlet.java:37]: I see. Originally, I was going to make this as a 13 line PR but I decided against it since it was so small. Now in the future, I'll keep note to add PRs like this.

File: capstone/src/main/webapp/dashboard-loader.js (hunk 9 above)
Comment (AndrewDPham) [reply capstone/src/main/webapp/dashboard-loader.js]: Got it.

File: capstone/src/main/java/com/google/sps/data/Room.java (hunk 6 above)
Comment (AndrewDPham) [reply capstone/src/main/java/com/google/sps/data/Room.java:63]: Got it. Thanks to @kyleasmith1  for explaining it more clearly to me.

**Review by 707090 — APPROVED** [review]
